[Goh Siau Wee] iP

weesiau opened the pull request:
### Duke: Power Hour
Task management system to help the productive person go **HARD**.

- text-based
- powerful, flexible
- constantly improving

Quick to pick up, easy to use. Just follow these simple steps!

1. Download .jar file from [here](https://github.com/weesiau/ip/releases/tag/v0.2)
2. Run/open the .jar file
3. Add and manage tasks with commands such as _todo, mark, deadline, list_ and more...
4. Enjoy increased productivity!

What is Duke capable of?

- [x] Add tasks
- [x] Manage tasks
- [ ] Display in GUI
- [ ] Have a personality
`and >`!

But what are **YOU** capable of? 🤔 

> You miss 100% of the shots you don't take. -Wayne Gretzky

Show us what you can do. And remember,
```java
public static void main(String[] args) {
    System.out.println("It's power hour!");
}
```

Commit:
Add Gradle support

weesiau committed:
Completed Level-1 Increment

Implemented skeletal version of the Duke chatbot with capability of echoing user's inputs and exiting when command is input

weesiau committed:
Completed Level-2 Increment

Added the ability to store user's input and display when command is given

weesiau committed:
Completed Level-3 Increment

Created new class, Task.java to encapsulate a task to be added and managed. Implemented ability to mark  added tasks as done or undone, and maintain the state of the tasks.

weesiau committed:
Completed Level-4 increment

Implemented support for the Event, Deadline and Todo Task types.

weesiau committed:
Completed A-TextUiTesting Increment

Implemented the semi-automated testing of Duke commands.

weesiau committed:
Completed Level-5 Increment

Implemented exception throwing and handling for major errors.

weesiau committed:
Completed Level-6 Increment

Implemented ability to delete a task from the list.

weesiau committed:
Completed A-Enums Increment

Altered program to use Enums to process user input.

weesiau committed:
Improved A-Enums Increment

Altered behaviour of program detecting illegal user input

weesiau committed:
Implement Level-7 increment

Implemented ability for Duke program to write tasks to a file on disk

weesiau committed:
Implement Level-8 increment

Improved handling of date and time deadlines for Deadline and Event tasks

weesiau committed:
Update Level-7 increment

weesiau committed:
Merge branch 'branch-Level-8'

weesiau committed:
Merge branch 'branch-Level-7'

weesiau committed:
Implement A-MoreOOP increment

Modified and refactored program to follow OOP paradigms.

weesiau committed:
Implement A-Packages increment

Packaged types into the duke package

weesiau committed:
Merge remote-tracking branch 'upstream/add-gradle-support'

weesiau committed:
Implemented A-Gradle increment

Integrated Gradle into the project

weesiau committed:
Implement A-JUnit increment

Implemented unit testing with gradle on Parser, TaskList and Task classes

weesiau committed:
Implement A-Jar increment

Packaged current program into a JAR file

weesiau committed:
Implement A-JavaDoc increment

Added javadoc comments to public classes and methods.

weesiau committed:
Implement A-CodingStandard increment

Edited source code to abide by coding standards

weesiau committed:
Implement Level-9 increment

Added ability for user to search for tasks by providing a search term as input

weesiau committed:
Add JavaDocs to public methods

weesiau committed:
Improve coding standard in source code

weesiau committed:
Update TaskList.java

weesiau committed:
Merge branch 'branch-A-CodingStandard'

weesiau committed:
Merge branch 'branch-Level-9'

Bacon-Strips commented on .gitignore:
```diff
@@ -15,3 +15,6 @@ bin/
 /text-ui-test/ACTUAL.txt
 text-ui-test/EXPECTED-UNIX.TXT
+*.class
+duke.jar
```
It is good that you have set these files to be ignored.

Bacon-Strips commented on .gitignore:
```diff
@@ -15,3 +15,6 @@ bin/
 
 /text-ui-test/ACTUAL.txt
 text-ui-test/EXPECTED-UNIX.TXT
+*.class
+duke.jar
+data/duketasks.txt
```
Good that you had added these files to be ignored by git.

Bacon-Strips commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,40 @@
+ */
+public class Deadline extends Task {
+
+    private final DateTimeFormatter FORMAT_TIME_YYYYMMDD = 
+            DateTimeFormatter.ofPattern("HH:mm 'on' yyyy/MM/dd");
```
```suggestion
    private static final DateTimeFormatter FORMAT_TIME_YYYYMMDD = 
            DateTimeFormatter.ofPattern("HH:mm 'on' yyyy/MM/dd");
```

It would be more appropriate to have this constant be a `static` variable.

Bacon-Strips commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,40 @@
+    private final DateTimeFormatter FORMAT_TIME_YYYYMMDD = 
+            DateTimeFormatter.ofPattern("HH:mm 'on' yyyy/MM/dd");
+            
+    private LocalDateTime deadLine;
```
```suggestion
    private final LocalDateTime deadLine;
```
Feel like it would be appropriate to set the attribute as `final`,

Bacon-Strips commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,99 @@
+ */
+public class Duke {
+
+    private Ui ui;
+
+    private Parser parser;
+
+    private Storage storage;
+
+    private TaskList taskList;
```
```suggestion
    private final Ui ui;

    private final Parser parser;

    private final Storage storage;

    private final TaskList taskList;
```
Suggesting the addition of the `final` modifier.

Bacon-Strips commented on src/main/java/duke/Event.java:
```diff
@@ -0,0 +1,40 @@
+ */
+public class Event extends Task {
+
+    private final DateTimeFormatter FORMAT_MMDDYY_TIME = 
+            DateTimeFormatter.ofPattern("MM/dd/yy, HH:mm");
+            
+    private LocalDateTime eventTime;
```
```suggestion
    private static final DateTimeFormatter FORMAT_MMDDYY_TIME = 
            DateTimeFormatter.ofPattern("MM/dd/yy, HH:mm");

    private final LocalDateTime eventTime;
```

Suggesting the addition of a few modifiers.

Bacon-Strips commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,70 @@
+ */
+public class Storage {
+
+    private File saveFile;
```
```suggestion
    private final File saveFile;
```
Suggestion of a `final` modifier.

Bacon-Strips commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,76 @@
+ * @author Siau Wee
+ *
+ */
+public class Task {
```
```suggestion
public abstract class Task {
```
Suggestion of the `abstract` modifier as the Task is not directly created in Duke.

Bacon-Strips commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,76 @@
+public class Task {
+
+    /** The name of the task. */
+    private String taskName;
```
```suggestion
    private final String taskName;
```
Suggesting addition of `final` modifier.

Bacon-Strips commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,130 @@
+ */
+public class TaskList {
+
+    private ArrayList<Task> addedTasks;
+
+    private Storage storage;
+
+    private Ui ui;
```
```suggestion
    private final ArrayList<Task> addedTasks;

    private final Storage storage;

    private final Ui ui;
```

Suggesting the addition of `final` modifier.

Bacon-Strips reviewed (commented):
Overall your code looks good, just suggesting the addition of a few `final` modifiers,

weesiau committed:
Implement GUI to replace CLI for Duke

weesiau committed:
Include assertions for key points in the program

There are some points in the program's lifetime where certain
assumptions should always hold true.

Assertions help codify requirements that render a program's behaviour
to be correct or not.

To better document these requirements, we assert certain conditions
to let programmers know that these conditions hold true
at these points in the program.

weesiau committed:
Refactor code handling program exit

Initial code for delaying the exit of the program, and displaying
and exit message was unclear and not abstract.

Refactor this portion of code to improve its quality, abiding by SLAP
and increasing code clarity.

weesiau committed:
Merge pull request #1 from weesiau/branch-A-Assertions

Include assertions for key points in the program

weesiau committed:
Merge branch 'branch-A-CodeQuality'

weesiau committed:
Implement ability to sort tasks by date/name in ascending order

Tasks should be able to be sorted by date and task name for better readability and usefulness.

Implemented the ability to sort the task list in ascending order by either the tasks dates or task names.

weesiau committed:
Implement loading of tasks from save file

weesiau committed:
Fix array index bugs and add product screenshot

weesiau committed:
Update README.md

weesiau committed:
Update README.md

weesiau committed:
Update test cases
